### PR TITLE
Squeeze before reduce for Welford

### PR DESCRIPTION
**Pull request by zasdfgbnm**
Fixes #ISSUE_NUMBER

### CODE REVIEW

File: torch/csrc/jit/codegen/cuda/arith.cpp [hunk 1]
```diff
@@ -1684,22 +1688,113 @@ WelfordResult Welford(
...
+  if (!reduction_axes.empty()) {
+    return WelfordRaw(squeezed, reduction_axes, init_avg, init_var, init_N);
+  }
+
+  // if squeeze only
+
+  if (init_N == nullptr) {
+    init_N = FusionGuard::getCurFusion()->zeroVal();
+  }
+  TensorView* out_N = full_like(squeezed, init_N, DataType::Index);
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/arith.cpp]: So, when `init_N` is null (I don't know if we would ever need to support non-null `init_N`), `out_N` is zero. When all reductions are trivial, shouldn't the out N be `1`?

Comment (zasdfgbnm) [reply torch/csrc/jit/codegen/cuda/arith.cpp]: I think what you said makes sense.

And I don't think non-null `init_N` is supported. Looks like non-zero init N is only useful to continue a welford from another welford, but this function only has one input.

Comment (naoyam) [reply torch/csrc/jit/codegen/cuda/arith.cpp]: `WelfordOp` itself should support non-null `init_N`. When an initial `WelfordOp` is rfactored, the second `WelfrodOp` takes the output `N` from the first `WelfordOp`. But, here we're defining an interface for the user, and in that case I don't know `init_N` can be non-null. Anyway, that's a different topic.

Comment (zasdfgbnm) [reply torch/csrc/jit/codegen/cuda/arith.cpp]: This is the code for rFactor of Welford:

```C++
    // Setup dependency chain, inserting producer before this op.
    // Expr* producer_definition =
    IrBuilder::create<WelfordOp>(
        producer_avg,
        producer_var,
        producer_n,
        wop->inAvg(),
        wop->inVar(),
        wop->inN(),
        wop->initAvg(),
        wop->initVar(),
        wop->initN());

    // Expr* consumer_definition =
    IrBuilder::create<WelfordOp>(
        wop->outAvg(),
        wop->outVar(),
        wop->outN(),
        producer_avg,
        producer_var,
        producer_n,
        wop->initAvg(),
        wop->initVar(),
        wop->initN());
```
Looks like we are using the same `initAvg`, `initVar` and `initN` for both `WelfordOp`. I don't know if this makes any sense. Shouldn't the second `WelfordOp` be totally ignoring these initial values?

Comment (naoyam) [reply torch/csrc/jit/codegen/cuda/arith.cpp]: Sorry, my bad. The output `N` of the first `WelfordOp` should be the input `N` of the second `WelfordOp`. The init values in the second `WelfordOp` shouldn't be needed. What actually happens is the predicate of the second `WelfordOp` is just always `true` as the inputs are at least initialized, so the init values are never used in the runtime Welford functions.

Comment (zasdfgbnm) [reply torch/csrc/jit/codegen/cuda/arith.cpp]: I have fixed this and added a new test `FusionSqueezeOnlyWelford_CUDA` for it

**Review by naoyam — APPROVED** [review]
LGTM

**Pull request merged**